File: app/frontend/types/borrower.ts
export interface BorrowerSummary {
  id: number
  name: string
  email: string | null
  phone: string | null
  archived: boolean
  loan_count: number
  total_principal: number
  total_interest_received: number
  last_payment_date: string | null
}

export interface BorrowerDetail {
  id: number
  name: string
  email: string | null
  phone: string | null
  address_line1: string | null
  address_line2: string | null
  city: string | null
  state: string | null
  zip: string | null
  notes: string | null
  tin: string | null
  archived: boolean
  created_at: string
}

File: app/models/borrower.rb
class Borrower < ApplicationRecord
  belongs_to :user
  has_many :loans, dependent: :restrict_with_error

  encrypts :tin

  validates :name, presence: true
  validates :user_id, presence: true
  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, allow_blank: true

  validate :tin_format, if: -> { tin.present? }

  scope :active, -> { where(archived: false) }
  scope :archived, -> { where(archived: true) }

  before_validation :normalize_tin

  def tin_masked
    return nil if tin.blank?
    "•••-••-#{tin.last(4)}"
  end

  def full_address
    [address_line1, city, state, zip].select(&:present?).join(", ")
  end

  def archive!
    update!(archived: true)
  end

  def unarchive!
    update!(archived: false)
  end

  private

  def normalize_tin
    self.tin = tin.gsub(/\D/, "") if tin.present?
  end

  def tin_format
    unless tin.match?(/\A\d{9}\z/)
      errors.add(:tin, "must be 9 digits (SSN or EIN)")
    end
  end
end
